File: test-validation/python/path-traversal/vulnerable.py
"""
漏洞示例：路径穿越（Path Traversal）- Python

问题说明：
直接使用用户输入构造文件路径，未做任何校验，
攻击者可以通过 "../" 等路径遍历字符访问系统任意文件。

预期检出：
- 行号：18, 27
- 规则 ID：priv-python-open-user-input
- 严重级别：ERROR
"""

import os

BASE_DIR = "/var/data/uploads"


def read_file_unsafe(filename: str) -> str:
    """漏洞：直接使用用户输入构造文件路径"""
    # 漏洞：直接拼接用户输入，未校验路径穿越
    filepath = os.path.join(BASE_DIR, filename)  # 第 19 行 - 路径穿越漏洞点
    with open(filepath, "r") as f:  # 第 20 行 - 打开未校验的文件
        return f.read()


def read_file_weak_filter(filename: str) -> str:
    """漏洞：仅简单替换 '../'，可被绕过"""
    # 漏洞：仅过滤 "../"，不过滤 "..\\" 或 URL 编码
    sanitized = filename.replace("../", "")  # 第 28 行 - 不完整的过滤
    filepath = os.path.join(BASE_DIR, sanitized)  # 第 29 行 - 仍可被绕过
    with open(filepath, "r") as f:
        return f.read()


def write_file_unsafe(filename: str, content: str) -> None:
    """漏洞：写入文件时未校验路径"""
    # 漏洞：直接拼接用户输入写文件
    filepath = os.path.join(BASE_DIR, filename)  # 第 37 行 - 路径穿越漏洞点
    with open(filepath, "w") as f:
        f.write(content)

File: test-validation/java/path-traversal/Vulnerable.java
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;

/**
 * 漏洞示例：路径穿越（Path Traversal）
 *
 * 问题说明：
 * 直接使用用户输入的文件名构造文件路径，未做任何校验，
 * 攻击者可以通过 "../" 等路径遍历字符访问系统任意文件。
 *
 * 预期检出：
 * - 行号：22
 * - 规则 ID：path-traversal-java-file
 * - 严重级别：ERROR
 */
public class Vulnerable {

    private static final String BASE_DIR = "/var/data/uploads";

    /**
     * 漏洞：直接使用用户输入构造文件路径，未校验路径穿越
     */
    public InputStream readFile(String fileName) throws IOException {
        // 漏洞：直接拼接用户输入，未过滤 "../" 等路径穿越字符
        File file = new File(BASE_DIR, fileName);  // 第 24 行 - 路径穿越漏洞点
        return new FileInputStream(file);
    }

    /**
     * 漏洞：使用 replace 过滤但不完整，可被绕过
     */
    public InputStream readFileWeakFilter(String fileName) throws IOException {
        // 漏洞：仅过滤 "../" 但不过滤 "..\" 或 URL 编码的 "../"
        String sanitized = fileName.replace("../", "");  // 第 32 行 - 不完整的过滤
        File file = new File(BASE_DIR, sanitized);  // 第 33 行 - 仍可被绕过
        return new FileInputStream(file);
    }
}
